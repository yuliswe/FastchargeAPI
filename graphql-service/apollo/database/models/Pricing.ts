import { PricingAvailability } from "@/__generated__/gql/graphql";
import dynamoose from "dynamoose";
import { Item } from "dynamoose/dist/Item";
import { PK, String_Required_NotEmpty, defaultCreatedAt, tableConfigs } from "../utils";

export { PricingAvailability };

export const PricingTableSchema = new dynamoose.Schema(
  {
    app: { hashKey: true, ...String_Required_NotEmpty("app") },
    createdAt: {
      rangeKey: true,
      type: Number,
      default: defaultCreatedAt,
    },
    name: { ...String_Required_NotEmpty("name") },
    callToAction: { type: String, default: "" },
    minMonthlyCharge: { type: String, default: "0" },
    chargePerRequest: { type: String, default: "0" },
    freeQuota: { type: Number, default: 0 },
    minMonthlyChargeFloat: Number,
    chargePerRequestFloat: Number,
    availability: {
      type: String,
      default: PricingAvailability.Public,
      enum: Object.values(PricingAvailability),
    },
  },
  {
    timestamps: {
      updatedAt: {
        updatedAt: {
          type: Number,
        },
      },
    },
  }
);
/// When creating a new Item class, remember to add it to codegen.yml mappers
/// config.

/**
 * Represents a pricing plan for an app.
 * @deletable false
 */
export class Pricing extends Item {
  app: PK;
  createdAt: number;
  name: string;
  callToAction: string;
  minMonthlyCharge: string;
  chargePerRequest: string;
  freeQuota: number;
  minMonthlyChargeFloat: number;
  chargePerRequestFloat: number;
  availability: PricingAvailability;
  updatedAt: number;
}

export type PricingCreateProps = {
  app: PK;
  name: string;
} & Partial<Pricing>;

export const PricingModel = dynamoose.model<Pricing>("Pricing", PricingTableSchema, { ...tableConfigs });
