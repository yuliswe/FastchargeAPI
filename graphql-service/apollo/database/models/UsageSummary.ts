import dynamoose from "dynamoose";
import { Item } from "dynamoose/dist/Item";
import { UsageSummaryStatus } from "../../__generated__/resolvers-types";
import { tableConfigs } from "../dynamodb";
import { PK, String_Required_NotEmpty, defaultCreatedAt } from "../utils";

/// When creating a new Item class, remember to add it to codegen.yml mappers
/// config.
/**
 * Represents a summary of API request access logs. This is used to calculate
 * the billing. Each UsageSummary item represents a single billing period and
 * can be billed to create AccountActivity item.
 */

export class UsageSummary extends Item {
  app: PK;
  subscriber: PK;
  /**
   * Used as a multiplier for charging different amount on an API call.
   * @default: 1
   */
  volume: number;
  createdAt: number;
  /** billed when account activities have been created */
  status: UsageSummaryStatus;
  billedAt: number | null;
  /** Number of usage logs in the queue when collected */
  numberOfLogs: number;
  /** ID of the AccountActivity item or null if not yet billed */
  billingRequestChargeAccountActivity: PK | null;
  /** ID of the AccountActivity item or null if not yet billed */
  billingMonthlyChargeAccountActivity: PK | null;
  /** ID of the AccountActivity item or null if not yet billed */
  appOwnerServiceFeeAccountActivity: PK | null;
  /** ID of the AccountActivity item or null if not yet billed */
  appOwnerRequestChargeAccountActivity: PK | null;
  /** ID of the AccountActivity item or null if not yet billed */
  appOwnerMonthlyChargeAccountActivity: PK | null;
  pricing: PK;
}

export type UsageSummaryCreateProps = {
  subscriber: PK;
  app: PK;
  pricing: PK;
  numberOfLogs: number;
} & Partial<UsageSummary>;

export const UsageSummaryTableSchema = new dynamoose.Schema(
  {
    subscriber: {
      hashKey: true,
      ...String_Required_NotEmpty("subscriber"),
    },
    createdAt: {
      type: Number,
      rangeKey: true,
      default: defaultCreatedAt,
    },
    app: {
      ...String_Required_NotEmpty("app"),
    },
    volume: { type: Number, default: 1 },
    status: {
      type: String,
      enum: Object.values(UsageSummaryStatus),
      default: UsageSummaryStatus.Pending,
    },
    billedAt: { type: Number, default: undefined },
    numberOfLogs: { type: Number, required: true },
    billingRequestChargeAccountActivity: { type: String, default: undefined },
    billingMonthlyChargeAccountActivity: { type: String, default: undefined },
    appOwnerServiceFeeAccountActivity: { type: String, default: undefined },
    appOwnerRequestChargeAccountActivity: { type: String, default: undefined },
    appOwnerMonthlyChargeAccountActivity: { type: String, default: undefined },
    pricing: { type: String, required: true },
  },
  {
    timestamps: {
      updatedAt: {
        updatedAt: {
          type: Number,
        },
      },
    },
  }
);
export const UsageSummaryModel = dynamoose.model<UsageSummary>("UsageSummary", UsageSummaryTableSchema, {
  ...tableConfigs,
});
