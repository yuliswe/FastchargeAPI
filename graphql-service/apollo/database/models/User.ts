import dynamoose from "dynamoose";
import { Item } from "dynamoose/dist/Item";
import { tableConfigs, validateStringDecimal } from "../utils";

/// When creating a new Item class, remember to add it to codegen.yml mappers
/// config.

export class User extends Item {
  uid: string;
  email: string;
  author: string;
  balance: string;
  balanceLimit: string;
  stripeCustomerId: string | null;
  stripeConnectAccountId: string | null;
  appTokens: { [appName: string]: string };
  createdAt: number;
  updatedAt: number;
}

export type UserCreateProps = {
  uid: string;
  email: string;
} & Partial<User>;

export enum UserTableIndex {
  IndexByEmailOnlyPk = "indexByEmail__onlyPK",
}

export const UserTableSchema = new dynamoose.Schema(
  {
    uid: { type: String, hashKey: true },
    email: {
      type: String,
      required: true,
      index: {
        type: "global",
        name: UserTableIndex.IndexByEmailOnlyPk,
        project: ["uid"],
      },
    },
    author: { type: String, default: () => `user_${Math.floor(Math.random() * 10000000)}` },
    stripeCustomerId: { type: String, required: false }, // Available after the user first tops up their account
    stripeConnectAccountId: { type: String, required: false }, // Available after the user first onboards their Stripe account
    balanceLimit: { type: String, default: "100", validate: validateStringDecimal("accountLimit") },
  },
  {
    timestamps: true,
  }
);

export const UserModel = dynamoose.model<User>("User", UserTableSchema, {
  ...tableConfigs,
});
