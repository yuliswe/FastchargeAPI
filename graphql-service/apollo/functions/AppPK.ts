export type AppPKContent = {
    name: string;
};

export class AppPK {
    static parse(pk: string): AppPKContent {
        return {
            name: pk,
        };
    }

    static stringify(item: AppPKContent): string {
        return item.name;
    }
}
