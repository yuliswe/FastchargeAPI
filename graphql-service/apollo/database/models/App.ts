import { AppVisibility } from "@/__generated__/gql/graphql";
import { GatewayMode } from "@/__generated__/resolvers-types";
import { validateAppName } from "@/functions/app";
import dynamoose from "dynamoose";
import { Item, tableConfigs } from "../utils";

export { AppVisibility, GatewayMode };

export const AppTableSchema = new dynamoose.Schema(
    {
        name: {
            hashKey: true,
            type: String,
            required: true,
            validate: validateAppName,
        },
        owner: {
            type: String,
            required: true,
            index: {
                type: "global",
                name: "indexByOwner__onlyPK",
                project: ["name", "owner"],
            },
        },
        logo: { type: String },
        title: { type: String },
        gatewayMode: { type: String, default: "proxy", enum: ["proxy", "redirect"] },
        description: { type: String },
        repository: { type: String },
        homepage: { type: String },
        readme: { type: String },
        visibility: { type: String, default: "private", enum: ["public", "private"] },

        deleted: { type: Boolean, default: false },
        deletedAt: { type: Number, required: false },
    },
    { timestamps: true }
);
/// When creating a new Item class, remember to add it to codegen.yml mappers
/// config.
export class App extends Item {
    name: string;
    owner: string;
    title: string;
    description: string;
    gatewayMode: GatewayMode;
    repository: string;
    homepage: string;
    readme: string;
    updatedAt: number;
    createdAt: number;
    visibility: AppVisibility;
    logo: string;
    deleted: boolean;
    deletedAt: number | null;
}

export type AppCreateProps = {
    name: string;
    owner: string;
} & Partial<App>;

export const AppModel = dynamoose.model<App>("App", AppTableSchema, {
    ...tableConfigs,
});
