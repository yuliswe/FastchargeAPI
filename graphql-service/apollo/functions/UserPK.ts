import { User } from "../dynamoose/models";

export class UserPK {
    static parse(pk: string): { email: string } {
        return {
            email: pk,
        };
    }

    static stringify(item: User): string {
        return item.email;
    }
}
