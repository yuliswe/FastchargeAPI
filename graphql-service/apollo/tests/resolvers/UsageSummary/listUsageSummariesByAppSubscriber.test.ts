import { graphql } from "@/__generated__/gql";
import { App } from "@/database/models/App";
import { Pricing } from "@/database/models/Pricing";
import { UsageSummary } from "@/database/models/UsageSummary";
import { User } from "@/database/models/User";
import { Can } from "@/permissions";
import { AppPK } from "@/pks/AppPK";
import { PricingPK } from "@/pks/PricingPK";
import { UserPK } from "@/pks/UserPK";
import { createTestApp } from "@/tests/test-data/App";
import { createTestPricing } from "@/tests/test-data/Pricing";
import { createTestUsageSummary } from "@/tests/test-data/UsageSummary";
import { createTestUser } from "@/tests/test-data/User";
import { baseRequestContext as context, simplifyGraphQLPromiseRejection } from "@/tests/test-utils/test-utils";
import { getTestGQLClient } from "@/tests/test-utils/testGQLClients";
import { beforeEach, describe, expect, jest, test } from "@jest/globals";

describe("listUsageSummariesByAppSubscriber", () => {
  let testSubscriber: User;
  let testOtherUser: User;
  let testAppOwner: User;
  let testApp: App;
  let testPricing: Pricing;
  const testUsageSummaries: UsageSummary[] = [];

  const listUsageSummariesByAppSubscriberQuery = graphql(`
    query TestListUsageSummary($subscriber: ID!, $app: ID!) {
      listUsageSummariesByAppSubscriber(subscriber: $subscriber, app: $app) {
        pk
        status
        billedAt
        billed
        volume
        app {
          pk
        }
        subscriber {
          pk
        }
        billingRequestChargeAccountActivity {
          pk
        }
      }
    }
  `);

  beforeEach(async () => {
    testOtherUser = await createTestUser(context);
    testSubscriber = await createTestUser(context);
    testAppOwner = await createTestUser(context);
    testApp = await createTestApp(context, {
      owner: UserPK.stringify(testAppOwner),
    });
    testPricing = await createTestPricing(context, {
      app: AppPK.stringify(testApp),
    });
    for (let i = 0; i < 3; i++) {
      const usageSummary = await createTestUsageSummary(context, {
        app: AppPK.stringify(testApp),
        subscriber: UserPK.stringify(testSubscriber),
        volume: 1,
        numberOfLogs: 1,
        pricing: PricingPK.stringify(testPricing),
      });
      testUsageSummaries.push(usageSummary);
    }
  });

  test("UsageSummary owner (subscriber) can list UsageSummaries", async () => {
    const result = await getTestGQLClient({ user: testSubscriber }).query({
      query: listUsageSummariesByAppSubscriberQuery,
      variables: {
        subscriber: UserPK.stringify(testSubscriber),
        app: AppPK.stringify(testApp),
      },
    });
    for (const usageSummary of result.data.listUsageSummariesByAppSubscriber) {
      expect(usageSummary).toMatchSnapshotExceptForProps({
        pk: expect.any(String),
        app: {
          pk: AppPK.stringify(testApp),
        },
        subscriber: {
          pk: UserPK.stringify(testSubscriber),
        },
      });
    }
  });

  test("A user cannot list UsageSummaries that they don't own", async () => {
    const result = getTestGQLClient({ user: testAppOwner }).query({
      query: listUsageSummariesByAppSubscriberQuery,
      variables: {
        subscriber: UserPK.stringify(testSubscriber),
        app: AppPK.stringify(testApp),
      },
    });
    await expect(simplifyGraphQLPromiseRejection(result)).rejects.toMatchObject([
      {
        code: "PERMISSION_DENIED",
        message: "You do not have permission to perform this action.",
        path: "listUsageSummariesByAppSubscriber",
      },
    ]);
  });

  test("A user cannot see the private fields of UsageSummaries on an object that they don't own", async () => {
    // Forces the listUsageSummariesByAppSubscriber to pass so that we can test the private
    // fields.
    jest.spyOn(Can, "listUsageSummariesByAppSubscriber").mockImplementation(() => Promise.resolve(true));
    const result = getTestGQLClient({ user: testOtherUser }).query({
      query: listUsageSummariesByAppSubscriberQuery,
      variables: {
        subscriber: UserPK.stringify(testSubscriber),
        app: AppPK.stringify(testApp),
      },
    });
    await expect(simplifyGraphQLPromiseRejection(result)).rejects.toMatchSnapshot();
  });

  test("App ower cannot see any field of a subscriber's UsageSummary other than volume", async () => {
    // Forces the listUsageSummariesByAppSubscriber to pass so that we can test the private
    // fields.
    jest.spyOn(Can, "listUsageSummariesByAppSubscriber").mockImplementation(() => Promise.resolve(true));
    const result = getTestGQLClient({ user: testAppOwner }).query({
      query: listUsageSummariesByAppSubscriberQuery,
      variables: {
        subscriber: UserPK.stringify(testSubscriber),
        app: AppPK.stringify(testApp),
      },
    });
    await expect(simplifyGraphQLPromiseRejection(result)).rejects.toMatchSnapshot();
  });

  test("App owner can see the volume field of a subscriber's UsageSummary", async () => {
    // Forces the listUsageSummariesByAppSubscriber to pass so that we can test the private
    // fields.
    jest.spyOn(Can, "listUsageSummariesByAppSubscriber").mockImplementation(() => Promise.resolve(true));
    const result = getTestGQLClient({ user: testAppOwner }).query({
      query: graphql(`
        query TestAppOwnerCanReadVolume($subscriber: ID!, $app: ID!) {
          listUsageSummariesByAppSubscriber(subscriber: $subscriber, app: $app) {
            volume
          }
        }
      `),
      variables: {
        subscriber: UserPK.stringify(testSubscriber),
        app: AppPK.stringify(testApp),
      },
    });
    await expect(result).resolves.toMatchSnapshot();
  });
});
