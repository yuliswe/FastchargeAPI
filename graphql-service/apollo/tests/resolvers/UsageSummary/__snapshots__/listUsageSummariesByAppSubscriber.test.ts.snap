// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listUsageSummariesByAppSubscriber A user cannot see the private fields of UsageSummaries on an object that they don't own 1`] = `
[
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.app",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billed",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billingRequestChargeAccountActivity",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.pk",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.status",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.subscriber",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.volume",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.1.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.1.billingRequestChargeAccountActivity",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.2.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.2.billingRequestChargeAccountActivity",
  },
]
`;

exports[`listUsageSummariesByAppSubscriber App ower cannot see any field of a subscriber's UsageSummary other than volume 1`] = `
[
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.app",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billed",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.billingRequestChargeAccountActivity",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.pk",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.status",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.0.subscriber",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.1.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.1.billingRequestChargeAccountActivity",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.2.billedAt",
  },
  {
    "code": "PERMISSION_DENIED",
    "message": "You do not have permission to perform this action.",
    "path": "listUsageSummariesByAppSubscriber.2.billingRequestChargeAccountActivity",
  },
]
`;

exports[`listUsageSummariesByAppSubscriber App owner can see the volume field of a subscriber's UsageSummary 1`] = `
{
  "data": {
    "listUsageSummariesByAppSubscriber": [
      {
        "__typename": "UsageSummary",
        "volume": 1,
      },
      {
        "__typename": "UsageSummary",
        "volume": 1,
      },
      {
        "__typename": "UsageSummary",
        "volume": 1,
      },
    ],
  },
  "loading": false,
  "networkStatus": 7,
}
`;

exports[`listUsageSummariesByAppSubscriber UsageSummary owner (subscriber) can list UsageSummaries 1`] = `
{
  "__typename": "UsageSummary",
  "app": {
    "__typename": "App",
    "pk": Any<String>,
  },
  "billed": false,
  "billedAt": null,
  "billingRequestChargeAccountActivity": null,
  "pk": Any<String>,
  "status": "pending",
  "subscriber": {
    "__typename": "User",
    "pk": Any<String>,
  },
  "volume": 1,
}
`;

exports[`listUsageSummariesByAppSubscriber UsageSummary owner (subscriber) can list UsageSummaries 2`] = `
{
  "__typename": "UsageSummary",
  "app": {
    "__typename": "App",
    "pk": Any<String>,
  },
  "billed": false,
  "billedAt": null,
  "billingRequestChargeAccountActivity": null,
  "pk": Any<String>,
  "status": "pending",
  "subscriber": {
    "__typename": "User",
    "pk": Any<String>,
  },
  "volume": 1,
}
`;

exports[`listUsageSummariesByAppSubscriber UsageSummary owner (subscriber) can list UsageSummaries 3`] = `
{
  "__typename": "UsageSummary",
  "app": {
    "__typename": "App",
    "pk": Any<String>,
  },
  "billed": false,
  "billedAt": null,
  "billingRequestChargeAccountActivity": null,
  "pk": Any<String>,
  "status": "pending",
  "subscriber": {
    "__typename": "User",
    "pk": Any<String>,
  },
  "volume": 1,
}
`;
